CONTRIBUTING.md

```diff
@@ -0,0 +1,232 @@
+
+  Fix any StyleCop issues reported in changed code. When adding new projects to VFS4G, be sure that StyleCop is analyzing them as part of the build.
+
+- *Prefer explicit types to interfaces and implicitly typed variables*
+
+  Avoid the use of `var` (C#) and `auto` (C++), and prefer concrete/explicit types to interface (e.g. prefer `List` to `IList`).
+
+  The VFS4G codebase uses this approach because:
+
+    * Interfaces can hide the performance characteristics of their underlying type.  For example, an `IDictionary` could be a `SortedList` or a `Dictionary` (or several other data types).
+    * Interfaces can hide the thread safety (or lack thereof) of their underlying type.  For example, an `IDictionary` could be a `Dictionary` or a `ConcurrentDictionary`.
+    * Explicit types make these performance and thread safety characteristics explicit when reviewing code.
+    * VFS4G is not a public API and its components are always shipped together.  Develoepers are free to make API changes to VFS4G's public methods.
+
+- *Method names start with a verb (e.g. "GetProjectedFolderEntryData" rather than "ProjectedFolderEntryData")*
+
+  Starting with a verb in the name improves readability and helps ensure consistency with the rest of the VFS4G codebase.
+
+- *Aim to write self-commenting code.  When necessary, comments should give background needed to understand the code.*
+
+  Helpful (good) comment:
+
+  ```
+  // Order the folders in descending order so that we walk the tree from bottom up.
+  // Traversing the folders in this order:
+  //  1. Ensures child folders are deleted before their parents
+  //  2. Ensures that folders that have been deleted by git 
+  //     (but are still in the projection) are found before their
+  //     parent folder is re-expanded (only applies on platforms 
+  //     where EnumerationExpandsDirectories is true)
+  foreach (PlaceholderListDatabase.PlaceholderData folderPlaceholder in   placeholderFoldersListCopy.OrderByDescending(x => x.Path))
+  ```
+
+  Obvious (bad) comment:
+
+  ```
+  // Check if enumeration expands directories on the current platform
+  if (GVFSPlatform.Instance.KernelDriver.EnumerationExpandsDirectories)
+  ```
+
+- *Add new interfaces when it makes sense for the product, not simply for unit testing*
+
+  When a class needs to be mocked (or have a subset of its behavior mocked), prefer using virtual methods instead of adding a new interface.  VFS4G uses interfaces when multiple implementations of the interface exist in the product code.
+
+- *Check for `null` using the equality (`==`) and inequality (`!=`) operators rather than `is`*
+
+  A corollary to this guideline is that equality/inequality operators that break `null` checks should not be added (see [this post](https://stackoverflow.com/questions/40676426/what-is-the-difference-between-x-is-null-and-x-null) for an example).
+
+- *Use `nameof(...)` rather than hardcoded strings*
```

Comment:
Do we need to mention that in some cases a hard-coded string is preferred for reporting to make sure we don't have to update reports when the name changes?

Reply:
That's a good idea, thanks!